### DISCUSSION TITLE
Data Import - Suggestions

### DISCUSSION THREAD

**Discussion by shogunpurple** [Budibase/budibase#649] [Ideas]
Importing data into budibase is a key element that has been asked for by quite a few customers.

The question now is how data **should** be imported into budibase applications.

At a high level, data import makes sense in these cases:
- Create a new table from an already existing dataset by importing a CSV
- Import **more** data into an existing table using a CSV

### Creating a new table from already existing data
In this scenario, we would need to determine the schema by reading the headers of the CSV and make them all strings. I don't feel like it's worth trying to be more sophisticated than this.

### Adding "more" data to an existing table
In this scenario, we first need to check that the headers match the existing table headers. This could be tricky, given that the existing table headers may not be strings. I suggest we do not implement this and only allow people to create new tables from data imports.

### UX
On creation of a new table in the builder, you see this UI:
![Screenshot 2020-09-30 at 11 00 07](https://user-images.githubusercontent.com/11256663/94671692-1c0c6f80-030c-11eb-8192-07302da2d9f6.png)

Adding a "Import from CSV" button here seems like a reasonable solution. Once you click that you can then select a CSV from your machine and use it to create a new table after we have validated it.

- What other considerations should we make for data importing into budibase? 
- What makes a great data import experience into no-code tools?
- In which cases is CSV import not enough?

**Comment by joebudi** [comment]
I agree and think we should keep things simple for now. CSV import on the creation of a new table suffices. 

What other considerations should we make for data importing into budibase?
Feedback and clarity. If the import fails, why and where? We should present this in the UI.

What makes a great data import experience into no-code tools?
Speed and simplicity. 
User clicks import > adds CSV > The columns, along with column types (editable dropdown), and status of each column (success/error) are presented to the user > User clicks 'Complete Import' or changes a column/doc to resolve errors.

In which cases is CSV import not enough?
2-way syncing of data.

**Reply by jdmoyer82** [reply]
As a related question to the above, I do not see a way to bulk import users into the Users table. Is there not a way to do this?

**Comment by mjashanks** [comment]
If we support an initial bulk import, I think that we should also support a bulk delete. Often people will "Have a go". They will realise a mistake, and try again.

So - i agree that starting with "Create Table from CSV" only makes sense, as they can just delete the table.
